File: C++/22.括号生成.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution
{
public:
    void DFS(int n, int countLeft, int countRight, string str)
    {
        if (countLeft == n && countRight == n)
            ans.push_back(str);
        if (countLeft < n)
            DFS(n, countLeft + 1, countRight, str + "(");
        if (countRight < n && countLeft > countRight)
            DFS(n, countLeft, countRight + 1, str + ")");
    }
    vector<string> generateParenthesis(int n)
    {
        DFS(n, 0, 0, "");
        return ans;
    }

private:
    vector<string> ans;
};
// @lc code=end

File: JavaScript/22.括号生成.js
/*
 * @lc app=leetcode.cn id=22 lang=javascript
 *
 * [22] 括号生成
 */

// @lc code=start
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
  const result = [];
  backtrack("", 0, 0, n, result);
  return result;
};
function backtrack(current, openCount, closeCount, n, result) {
  if (openCount === n && closeCount === n) {
    result.push(current);
    return;
  }

  if (openCount < n) {
    backtrack(current + "(", openCount + 1, closeCount, n, result);
  }

  if (closeCount < openCount) {
    backtrack(current + ")", openCount, closeCount + 1, n, result);
  }
}
// @lc code=end
